File: lib/src/models/dynamic_marker_position_update.dart
/// Represents a position update for a dynamic marker.
///
/// These updates are typically received from an external data source
/// (WebSocket, Firebase, MQTT, etc.) and converted to this format
/// before being passed to the plugin.
class DynamicMarkerPositionUpdate {
  /// The marker ID this update applies to.
  final String markerId;

  /// New latitude coordinate.
  final double latitude;

  /// New longitude coordinate.
  final double longitude;

  /// New heading in degrees (0-360, north = 0).
  final double? heading;

  /// Current speed in meters per second.
  final double? speed;

  /// Altitude in meters (for 3D tracking scenarios).
  final double? altitude;

  /// GPS accuracy in meters.
  final double? accuracy;

  /// Timestamp when this position was recorded.
  final DateTime timestamp;

  /// Additional data associated with this update.
  final Map<String, dynamic>? additionalData;

  /// Creates a position update for a dynamic marker.
  const DynamicMarkerPositionUpdate({
    required this.markerId,
    required this.latitude,
    required this.longitude,
    required this.timestamp,
    this.heading,
    this.speed,
    this.altitude,
    this.accuracy,
    this.additionalData,
  });

  /// Creates an update from a JSON map.
  factory DynamicMarkerPositionUpdate.fromJson(Map<String, dynamic> json) {
    return DynamicMarkerPositionUpdate(
      markerId: json['markerId'] as String,
      latitude: (json['latitude'] as num).toDouble(),
      longitude: (json['longitude'] as num).toDouble(),
      timestamp: DateTime.parse(json['timestamp'] as String),
      heading:
          json['heading'] != null ? (json['heading'] as num).toDouble() : null,
      speed: json['speed'] != null ? (json['speed'] as num).toDouble() : null,
      altitude:
          json['altitude'] != null ? (json['altitude'] as num).toDouble() : null,
      accuracy:
          json['accuracy'] != null ? (json['accuracy'] as num).toDouble() : null,
      additionalData: json['additionalData'] != null
          ? Map<String, dynamic>.from(json['additionalData'] as Map)
          : null,
    );
  }

  /// Creates an update from a generic map.
  ///
  /// Useful for converting data from JSON/WebSocket messages.
  /// Supports various coordinate key names:
  /// - latitude/longitude
  /// - lat/lng
  /// - lat/lon
  factory DynamicMarkerPositionUpdate.fromMap(Map<String, dynamic> map) {
    // Extract latitude - support multiple key names
    final lat = map['latitude'] ?? map['lat'];
    final double latitude = lat is num ? lat.toDouble() : double.parse(lat.toString());

    // Extract longitude - support multiple key names
    final lng = map['longitude'] ?? map['lng'] ?? map['lon'];
    final double longitude = lng is num ? lng.toDouble() : double.parse(lng.toString());

    // Parse timestamp if provided
    DateTime timestamp;
    if (map['timestamp'] != null) {
      timestamp = map['timestamp'] is String
          ? DateTime.parse(map['timestamp'] as String)
          : map['timestamp'] as DateTime;
    } else {
      timestamp = DateTime.now();
    }

    return DynamicMarkerPositionUpdate(
      markerId: map['id'] as String,
      latitude: latitude,
      longitude: longitude,
      timestamp: timestamp,
      heading:
          map['heading'] != null ? (map['heading'] as num).toDouble() : null,
      speed: map['speed'] != null ? (map['speed'] as num).toDouble() : null,
      altitude:
          map['altitude'] != null ? (map['altitude'] as num).toDouble() : null,
      accuracy:
          map['accuracy'] != null ? (map['accuracy'] as num).toDouble() : null,
      additionalData: map['data'] != null
          ? Map<String, dynamic>.from(map['data'] as Map)
          : null,
    );
  }

  /// Converts this update to a JSON map.
  Map<String, dynamic> toJson() {
    return {
      'markerId': markerId,
      'latitude': latitude,
      'longitude': longitude,
      'timestamp': timestamp.toIso8601String(),
      'heading': heading,
      'speed': speed,
      'altitude': altitude,
      'accuracy': accuracy,
      'additionalData': additionalData,
    };
  }

  /// Creates a copy with updated fields.
  DynamicMarkerPositionUpdate copyWith({
    String? markerId,
    double? latitude,
    double? longitude,
    DateTime? timestamp,
    double? heading,
    double? speed,
    double? altitude,
    double? accuracy,
    Map<String, dynamic>? additionalData,
  }) {
    return DynamicMarkerPositionUpdate(
      markerId: markerId ?? this.markerId,
      latitude: latitude ?? this.latitude,
      longitude: longitude ?? this.longitude,
      timestamp: timestamp ?? this.timestamp,
      heading: heading ?? this.heading,
      speed: speed ?? this.speed,
      altitude: altitude ?? this.altitude,
      accuracy: accuracy ?? this.accuracy,
      additionalData: additionalData ?? this.additionalData,
    );
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is DynamicMarkerPositionUpdate &&
        other.markerId == markerId &&
        other.timestamp == timestamp;
  }

  @override
  int get hashCode => markerId.hashCode ^ timestamp.hashCode;

  @override
  String toString() {
    return 'DynamicMarkerPositionUpdate(markerId: $markerId, '
        'lat: $latitude, lng: $longitude, timestamp: $timestamp)';
  }
}

File: android/src/main/kotlin/com/neiladunlop/fluttermapboxnavigation2/models/DynamicMarkerPositionUpdate.kt
package com.neiladunlop.fluttermapboxnavigation2.models

/**
 * Represents a position update for a dynamic marker.
 *
 * These updates are typically received from an external data source
 * (WebSocket, Firebase, MQTT, etc.) and converted to this format
 * before being passed to the marker manager.
 */
data class DynamicMarkerPositionUpdate(
    /** The marker ID this update applies to. */
    val markerId: String,

    /** New latitude coordinate. */
    val latitude: Double,

    /** New longitude coordinate. */
    val longitude: Double,

    /** Timestamp when this position was recorded (ISO8601 string). */
    val timestamp: String,

    /** New heading in degrees (0-360, north = 0). */
    val heading: Double? = null,

    /** Current speed in meters per second. */
    val speed: Double? = null,

    /** Altitude in meters (for 3D tracking scenarios). */
    val altitude: Double? = null,

    /** GPS accuracy in meters. */
    val accuracy: Double? = null,

    /** Additional data associated with this update. */
    val additionalData: Map<String, Any>? = null
) {
    companion object {
        /**
         * Creates a DynamicMarkerPositionUpdate from a JSON map.
         */
        @Suppress("UNCHECKED_CAST")
        fun fromJson(json: Map<String, Any>): DynamicMarkerPositionUpdate {
            return DynamicMarkerPositionUpdate(
                markerId = json["markerId"] as String,
                latitude = (json["latitude"] as Number).toDouble(),
                longitude = (json["longitude"] as Number).toDouble(),
                timestamp = json["timestamp"] as String,
                heading = (json["heading"] as? Number)?.toDouble(),
                speed = (json["speed"] as? Number)?.toDouble(),
                altitude = (json["altitude"] as? Number)?.toDouble(),
                accuracy = (json["accuracy"] as? Number)?.toDouble(),
                additionalData = json["additionalData"] as? Map<String, Any>
            )
        }

        /**
         * Creates an update from a generic map with flexible key names.
         *
         * Supports various coordinate key names:
         * - latitude/longitude
         * - lat/lng
         * - lat/lon
         */
        @Suppress("UNCHECKED_CAST")
        fun fromMap(map: Map<String, Any>): DynamicMarkerPositionUpdate {
            // Extract latitude - support multiple key names
            val lat = map["latitude"] ?: map["lat"]
            val latitude = when (lat) {
                is Number -> lat.toDouble()
                is String -> lat.toDouble()
                else -> throw IllegalArgumentException("Missing latitude")
            }

            // Extract longitude - support multiple key names
            val lng = map["longitude"] ?: map["lng"] ?: map["lon"]
            val longitude = when (lng) {
                is Number -> lng.toDouble()
                is String -> lng.toDouble()
                else -> throw IllegalArgumentException("Missing longitude")
            }

            // Extract marker ID
            val markerId = (map["markerId"] ?: map["id"]) as String

            // Parse timestamp with fallback to current time
            val timestamp = map["timestamp"] as? String
                ?: java.time.Instant.now().toString()

            return DynamicMarkerPositionUpdate(
                markerId = markerId,
                latitude = latitude,
                longitude = longitude,
                timestamp = timestamp,
                heading = (map["heading"] as? Number)?.toDouble(),
                speed = (map["speed"] as? Number)?.toDouble(),
                altitude = (map["altitude"] as? Number)?.toDouble(),
                accuracy = (map["accuracy"] as? Number)?.toDouble(),
                additionalData = map["data"] as? Map<String, Any>
                    ?: map["additionalData"] as? Map<String, Any>
            )
        }
    }

    /**
     * Returns the position as a LatLng.
     */
    val position: LatLng
        get() = LatLng(latitude, longitude)

    /**
     * Converts this update to a JSON map.
     */
    fun toJson(): Map<String, Any?> {
        return mapOf(
            "markerId" to markerId,
            "latitude" to latitude,
            "longitude" to longitude,
            "timestamp" to timestamp,
            "heading" to heading,
            "speed" to speed,
            "altitude" to altitude,
            "accuracy" to accuracy,
            "additionalData" to additionalData
        )
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as DynamicMarkerPositionUpdate
        return markerId == other.markerId && timestamp == other.timestamp
    }

    override fun hashCode(): Int {
        return markerId.hashCode() * 31 + timestamp.hashCode()
    }

    override fun toString(): String {
        return "DynamicMarkerPositionUpdate(markerId='$markerId', " +
                "lat=$latitude, lng=$longitude, timestamp=$timestamp)"
    }
}
